### PR TITLE
[TRTLLM-8768][chore] Fuse QK down_proj with indexer K + weight_proj for FP4 ckpt

### PR DESCRIPTION
**Pull request by chang-l**
Perf implication:

Without this PR:
concurrency=64;  ISL/OSL=1K/2K;  DEP=8:
```
Request Throughput (req/sec):                     0.9469
Total Output Throughput (tokens/sec):             1939.3511
Total Token Throughput (tokens/sec):              2926.8559
Total Latency (ms):                               67585.4940
Average request latency (ms):                     67518.0704
Per User Output Throughput [w/ ctx] (tps/user):   30.3326
Per GPU Output Throughput (tps/gpu):              242.4189
```

With this PR (~1.02X):
concurrency=64;  ISL/OSL=1K/2K;  DEP=8:
```
Request Throughput (req/sec):                     0.9670
Total Output Throughput (tokens/sec):             1980.3425
Total Token Throughput (tokens/sec):              2988.7198
Total Latency (ms):                               66186.5319
Average request latency (ms):                     66121.9604
Per User Output Throughput [w/ ctx] (tps/user):   30.9731
Per GPU Output Throughput (tps/gpu):              247.5428
```

## Release Notes

* **New Features**
  * Added support for DeepseekV3.2 attention variant.

* **Refactor**
  * Optimized sparse attention backend with improved indexer-based attention pathway for enhanced performance.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py [hunk 1]
```diff
@@ -1085,65 +1101,95 @@ def sparse_attn_indexer(
...
         q_fp8 = q_fp8.view(-1, self.n_heads, self.head_dim)
         q_scale = q_scale.view(-1, self.n_heads, 1)
 
-        weights = self.weights_proj(hidden_states)
-        weights = weights.unsqueeze(
-            -1) * q_scale * self.softmax_scale * self.n_heads**-0.5
-        weights = weights.squeeze(-1)
+        # indexer weight scaling
+        # Lazy compilation on first call
+        if self._weight_scale_compiled is None:
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa.py]: Can we directly add `@torch.compile(dynamic=True, options={"max-autotune": True})` to `self.weight_scale`?

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py [hunk 2]
```diff
@@ -1085,65 +1101,95 @@ def sparse_attn_indexer(
 
         return topk_indices_buffer
 
+    def weight_scale(self, hidden_states: torch.Tensor,
+                     indexer_weights: Optional[torch.Tensor],
+                     q_scale: torch.Tensor) -> torch.Tensor:
+        weights = indexer_weights if indexer_weights is not None else self.weights_proj(
+            hidden_states)
+        weights = weights.unsqueeze(
+            -1) * q_scale * self.softmax_scale * self.n_heads**-0.5
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa.py]: How about changing the computation to `weights.unsqueeze(-1) * (q_scale * (self.softmax_scale * self.n_heads**-0.5))`? First scalar, then tensor. But I'm not sure whether it helps when using `torch.compile`.

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py [hunk 3]
```diff
@@ -1085,65 +1101,95 @@ def sparse_attn_indexer(
...
+            )
+        else:
+            q, k = maybe_execute_in_parallel(
+                lambda: self.wq_b(qr),
+                lambda: self.wk(hidden_states),
+                self.ln_events[0],
+                self.ln_events[1],
+                self.aux_stream,
+            )
+            k = self.k_norm(k)
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa.py]: Can we also run k_norm in parallel (i.e., use `lambda: self.k_norm(self.wk(hidden_states))` instead)?

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py (hunk 2 above)
Comment (chang-l) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa.py]: Yeah, with torch.compile, it should be the same — the profiler just shows a single triton kernel.

File: tensorrt_llm/_torch/models/modeling_deepseekv3.py [hunk 4]
```diff
@@ -544,6 +555,65 @@ def __init__(
...
+                             type=PositionEmbeddingType.yarn,
+                             rope=RopeParams.from_config(config),
+                             is_neox=False,
+                         ),
+                         layer_idx=layer_idx,
+                         dtype=config.torch_dtype,
+                         config=model_config,
+                         aux_stream=aux_stream)
+
+        if self.fuse_a_indexer_k_weight:
```
[the remark this answers is not in the document]
Comment (chang-l) [reply tensorrt_llm/_torch/models/modeling_deepseekv3.py]: good point -- updated.

**Review by yuxianq — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_deepseekv3.py [hunk 5]
```diff
@@ -544,6 +545,79 @@ def __init__(
...
+
+        self.indexer = self.mqa.indexer
+
+        if self.fuse_a_indexer_k_weight:
+            # For DeepseekV32, the kv_a_proj_with_mqa includes:
+            # q_a_proj + kv_a_proj_with_mqa + indexer.wk + indexer.weights_proj
+            self.kv_a_proj_with_mqa = DeepseekV3Linear(
+                config.hidden_size,
+                self.kv_lora_rank + self.qk_rope_head_dim + self.q_lora_rank +
+                self.indexer.head_dim + self.indexer.n_heads,
```
Comment (syuoni) [inline tensorrt_llm/_torch/models/modeling_deepseekv3.py]: If we change this module's output dimension here, does the weight loading still work?

File: tensorrt_llm/_torch/models/modeling_deepseekv3.py [hunk 6]
```diff
@@ -544,6 +545,79 @@ def __init__(
...
+                skip_create_weights_in_init=model_config.
+                skip_create_weights_in_init,
+                use_custom_cublas_mm=True)
+
+    def post_load_weights(self):
+        if self.fuse_a_indexer_k_weight:
+            assert self.kv_a_proj_with_mqa.weight.data.dtype == self.indexer.wk.weight.data.dtype == self.indexer.weights_proj.weight.data.dtype, "all weights in kv_a_proj_with_mqa module must have matching dtype"
+            # Copy indexer weights into the fused kv_a_proj_with_mqa module
+            indexer_wk_weight = self.indexer.wk.weight.data
+            offset = self.kv_lora_rank + self.qk_rope_head_dim + self.q_lora_rank
```
Comment (syuoni) [inline tensorrt_llm/_torch/models/modeling_deepseekv3.py]: Might be related to another thread. I don't see the weight loading for the first `self.kv_lora_rank + self.qk_rope_head_dim + self.q_lora_rank` rows, are they correctly handled?

If they are handled in another place, I'm wondering if we can place all the weight loading logics in a same place. That would reduce some confusion.

**Review by lfr-0531 — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_deepseekv3.py (hunk 5 above)
Comment (chang-l) [reply tensorrt_llm/_torch/models/modeling_deepseekv3.py]: Yes, when we change the module's output dim, it would change its weights shape too, more specifically, from `[self.kv_lora_rank + self.qk_rope_head_dim + self.q_lora_rank, hidden_size]` to `[self.kv_lora_rank + self.qk_rope_head_dim + self.q_lora_rank + self.indexer.head_dim + self.indexer.n_heads, hidden_size]`.  
It actually expands the weights in the row dimension (fortunately).  So, [here](https://github.com/chang-l/TensorRT-LLM/blob/faa571916270ede42aae737dac3792791c0d631f/tensorrt_llm/_torch/models/modeling_deepseekv3.py#L366-L368), we first load the original `kv_a_proj_with_mqa` weight to first chunk of rows; and then, [here](https://github.com/chang-l/TensorRT-LLM/blob/faa571916270ede42aae737dac3792791c0d631f/tensorrt_llm/_torch/models/modeling_deepseekv3.py#L609), we copy the indexer's weights to the successive chunk of rows.

File: tensorrt_llm/_torch/models/modeling_deepseekv3.py (hunk 6 above)
Comment (syuoni) [reply tensorrt_llm/_torch/models/modeling_deepseekv3.py]: Thanks for the explanation.

**Review by syuoni — APPROVED** [review]
